### DISCUSSION TITLE
Use ttb in ANSYS UPF

### DISCUSSION THREAD

**Discussion by edwardXZ06** [adtzlr/ttb#28] [General]
Hi:

I am developing a constitutive model on ANSYS, and it is known that ANSYS UPF also use Fortran for its user defined material. Even though the ttb has tested its application in Abaqus, I would still like to give it a try to see if it can work with ANSYS, especially considering ANSYS use the same intel fortran compiler like Abaqus. 

While I saved the ttb folder at the work directory, there is an error reported below:
''
usermat.F
usermat.F(154): error #6278: This USE statement is not positioned correctly within the scoping unit.
      use Tensor
------^
''
Based on this, how do you think the possibility to use ttb in ANSYS?

Thanks,

Xiang
![image](https://user-images.githubusercontent.com/88357649/170911591-33028a3a-a8f2-43d5-a5a0-8a9063fed3e9.png)

![image](https://user-images.githubusercontent.com/88357649/170911448-3dd82919-46ad-494e-9554-626f278e2e5d.png)

**Comment by jfriedlein** [comment]
Hello Xiang @edwardXZ06,

because the toolbox is made for Fortran in general, it should also work for Ansys. The subroutine usermat in the first picture looks quite promising, but please verify that internal conventions (ordering of components in Voigt notation, factor 2 on shear strains, etc.) are consistent with the toolbox or use the according function variants (e.g. "As Array" vs "As Abaqus Array" [ttb Module Function Overview](https://adtzlr.github.io/ttb/functions.html)).

Regarding your above error #6278, as the compiler hints the "use Tensor" statement is positioned incorrectly in the subroutine. "use Tensor" needs to be placed before all declarations and includes, so directly in a line after "subroutine .... ( ... )" (also shown [here](https://github.com/jfriedlein/usrmat_LS-Dyna_Fortran#our-first-user-material)). So, once you place "use Tensor" before the "#include "impcom.inc" and possibly earlier includes, the compiler should no longer complain.

Once you find some compatibility issues or make the ttb work in Ansys, please let us know and provide your operating system, Ansys version, Fortran compiler, etc. This lets us document compatibilities and will help other users.

Best regards,
Johannes
